---
title: "Rna-seq analysis"
author: "Polina Kozyulina"
date: "March 11, 2017"
output:
  html_document: default
  pdf_document: default
---

```{r setup, message=FALSE}
library(DESeq2)
library(ggplot2)
library(ggrepel)
library(clusterProfiler)
library(org.Hs.eg.db)
library(pheatmap)
library(dplyr)
library(VennDiagram)
options(width=140)
setwd('/media/polina/305622A3562269B4/bioR/rbio')
```

## Файлы с которыми будем работать сегодня

В ходе данной лекции нам понадобятся следующие файлы "GSE89225_illumina_counts.csv", "conditions.csv", "human_mart.txt". Для начала убедимся в том, что мы можем эти файлы прочитать. И посмотрим, что в них находится.

```{r load, cache=TRUE}
# reading data
counts <- read.csv("GSE89225_Illumina_counts.csv", row.names=1)
conditions <- read.csv("conditions.csv", row.names=1)
mart <- read.table("human_mart.txt", sep="\t", header=1, check.names = F)

# Getting rid of outlier
counts <- as_data_frame(select(counts, -treg_NBP_patient3))
conditions <- conditions[rownames(conditions) != "treg_NBP_patient3", ]
samples <- rownames(conditions)

```
# Sanity checks
Нужно всегда проверять длины библиотек и количество rRNA reads, которые оказались в библиотеке. Количество ридов можно проверять после выравнивания или после квантификации.

```{r sanity_check, cache=TRUE}
proteinCoding <- mart[mart[, 3] == "protein_coding", ]
rRNA <- mart[mart[, 3] == "rRNA", ]

pcCounts <- counts[rownames(counts) %in% as.character(proteinCoding[, 1]), ]
rrnaCounts <- counts[rownames(counts) %in% as.character(rRNA[, 1]), ]

sampleCount <- ncol(counts)
toPlot <- data.frame(
  sample=rep(colnames(counts), 3),
  value=c(colSums(counts) - colSums(pcCounts) - colSums(rrnaCounts), 
          colSums(pcCounts), 
          colSums(rrnaCounts)),
  type=c(rep("other", sampleCount), 
         rep("protein coding", sampleCount),
         rep("rrna", sampleCount))
)

plot <- ggplot(data=toPlot, aes(x=sample, y=value, fill=type)) +
  geom_bar(stat="identity") + theme_bw() + 
  theme(axis.text.x = element_text(angle=90, vjust=0.5))
plot

```


# Обработка данных пакетом DESeq2
DESeq2 -- это царь-библиотека необходимая для работы с данными RNA-seq. Тут и дифференциальная экспрессия, и нормализации, и PCA-plots.

```{r deseq_prep, cache=TRUE, message=FALSE}
## Cache True so it wouldn't redo this chunk every time you reload Rmd

# Tregs vs Tconv
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = conditions,
                              design =  ~ tissue + cells) # 
dds <- dds[rowSums(counts(dds)) > 20, ]
dds <- DESeq(dds)
vst_dds <- varianceStabilizingTransformation(dds) # variance stabilizing transformation
counts.norm <- assay(vst_dds)

# Breast tumor vs NBP
dds_hw <- DESeqDataSetFromMatrix(countData = counts,
                              colData = conditions,
                              design = ~ cells + tissue) # 
dds_hw <- dds_hw[rowSums(counts(dds_hw)) > 20, ]
dds_hw <- DESeq(dds_hw)
vst_dds_hw <- varianceStabilizingTransformation(dds_hw) # variance stabilizing transformation
counts.norm_hw <- assay(vst_dds_hw)

```
# Строим PCA для смопроверки

```{r PCA, cache=TRUE}
pca_data <- prcomp(t(counts.norm))
percents <- pca_data$sdev^2 / sum(pca_data$sdev^2)
to_plot <- t(counts.norm) %*% pca_data$rotation

gdata <- data.frame(
  x=to_plot[, 1],
  y=to_plot[, 2],
  tissue=conditions[, 1],
  cells=conditions[, 2],
  name=rownames(conditions)
)

ggplot(data=gdata, aes(x=x, y=y, color=cells, shape=tissue, text=name)) +
  geom_point(size=3) + theme_bw()  +
  xlab(paste0("PC", 1, ": ", formatC(100 * percents[1], digits=4), "%")) +
  ylab(paste0("PC", 2, ": ", formatC(100 * percents[2], digits=4), "%")) #+
  #geom_text_repel(data=gdata,aes(label=rownames(gdata)))

```

# Differential expression
Давайте посмотрим, как выглядят результаты дифференциальной экспрессии и отсортируем их по статистике.

```{r diff_exp}
# Tregs vs Tconv
res_pre <- results(dds)
mcols(res_pre)$description
res <- res_pre[order(res_pre[, 4]), ]
res

# BT vs NBP
res_hw_pre <- results(dds_hw)
mcols(res_hw_pre)$description
res_hw <- res_hw_pre[order(res_hw_pre[, 4]), ]
res_hw

```

# Volcano plot
Самый обычный способ визуализировать данные дифф.экспрессии -- это volcano plot. По оси x мы будем откладывать log fold change, а по y - adjusted p value.

```{r volcano_plot}

# Tregs vs Tconv
gdata <- data.frame(x=res$log2FoldChange, y=-log10(res$padj), exp = "Tregs vs Tconv")

# BT vs NBP
gdata_hw <- data.frame(x=res_hw$log2FoldChange, y=-log10(res_hw$padj), exp = "Breast tumor vs Normal breat parenhima")

# Combining
gdata_full <- as.data.frame(rbind(gdata, gdata_hw))
gdata_full <- gdata_full[complete.cases(gdata_full), ]
gdata_full$Significance <- ifelse(gdata_full$y > -log10(0.01), "Significant", "Not significant")
gdata_full$Significance <- as.factor(gdata_full$Significance)

ggplot(data=gdata_full, aes(x=x, y=y, col=Significance)) +
  geom_point(size=1) + theme_bw()  +
  xlab("Log fold change") +
  ylab("Adjusted p.value") +
  facet_grid(. ~ exp) +
  geom_hline(yintercept=-log10(0.01), col = 'red', linetype = 'dashed') +
  scale_color_manual(values = c('black', 'red'))
```

# Heatmap
Также мы можем построить тепловую карту отсортировав гены по статистике

```{r heatmap}
# getting genes from a pathway
kkeys <- keys(org.Hs.eg.db, keytype="ENSEMBL")
goAnno <- AnnotationDbi::select(org.Hs.eg.db, keys=kkeys, 
                                keytype="ENSEMBL", columns=c("GOALL", "ONTOLOGYALL", "SYMBOL"))
goAnno <- tbl_df(goAnno)
goAnno <- filter(goAnno, GOALL=="GO:0007159") # у меня этот путь почему-то не выделяется на фоне других?!?

# or you can pick ENTREZ, or SYMBOL, or whatever you want
ENSEMBL <- goAnno[!duplicated(goAnno$ENSEMBL), ]$ENSEMBL
genesToVisualise <- as.data.frame(ENSEMBL)
rownames(genesToVisualise) <- genesToVisualise[, 1]
nrow(genesToVisualise)

#counts.norm <- counts(dds, normalized=TRUE)


# visualizing... 

to_visualise <- as.data.frame(counts.norm[, order(conditions[, 2])]) #sorting order(res$stat)
to_visualise$stat <- res_pre$stat
to_visualise$ENSEMBL <- rownames(to_visualise)
to_visualise_merged <- merge(to_visualise, genesToVisualise) # getting only genes that we need
to_visualise_merged <- to_visualise_merged[order(to_visualise_merged$stat), ] # sorting by stat (even though it was sorted before)

to_visualise_merged$stat <- NULL #deleting unnessesary cols
to_visualise_merged$ENSEMBL <- NULL #deleting unnessesary cols
to_visualise_merged <- t(apply(to_visualise_merged, 1, function(r) {(r - min(r)) / (max(r) - min(r)) }))

pheatmap(to_visualise_merged, 
         show_rownames = F, cluster_rows = F,
         cluster_cols=F,
         annotation_col = conditions)
```

# VennDiagram
A diagram showing interseption of two diff expressing gene sets depending on grouping

```{r VennDiagram}
setA <- rownames(res[res$padj < 0.01 & !is.na(res$padj), ]) # & abs(res$log2FoldChange) > 1
setB <- rownames(res_hw[res_hw$padj < 0.01 & !is.na(res_hw$padj), ]) # & abs(res_hw$log2FoldChange) > 1
commonLetters <- intersect(setA, setB)

draw.pairwise.venn(length(setA), length(setB), length(commonLetters),
                   category = c("Treg vs Tconv", "TB vs NBP"),
                   fill = c("cyan", "violet"),
                   lty = "blank",
                   cat.pos = 1,
                   alpha = 0.3,
                   cat.dist = 0.05)

```





